listenbrainz_spark/ftp/download.py

```diff
@@ -0,0 +1,133 @@
+        """
+        self.connection.cwd('/pub/musicbrainz/listenbrainz/labs/mappings/msid-mbid-mapping/')
+        mappings_dump = self.list_dir()
+        req_mapping_dump = self.get_req_dump(mappings_dump, mapping_dump_id, 3)
+
+        self.connection.cwd(req_mapping_dump)
+        mapping_file_name = self.get_file_name(req_mapping_dump)
+
+        t0 = time()
+        current_app.logger.info('Downloading {} from FTP...'.format(mapping_file_name))
+        dest_path = self.download_dump(mapping_file_name, directory)
+        current_app.logger.info('Done. Total time: {:.2f} sec'.format(time() - t0))
+        return dest_path
+
+    def download_msid_mbid_mapping_with_text(self):
+        pass
+
+    def download_msid_mbid_mapping_with_matchable(self):
+        pass
+
+    def download_listens(self, directory, listens_dump_id=None):
+        """ Download listens to dir passed as an argument.
+
+            Args:
+                directory (str): Dir to save listens locally.
+                listens_dump_id (int): Unique identifier of listens to be downloaded.
+                    If not provided, most recent listens will be downloaded.
+
+            Returns:
+                dest_path (str): Local path where listens have been downloaded.
+        """
+        self.connection.cwd('/pub/musicbrainz/listenbrainz/fullexport/')
+        listens_dump = self.list_dir()
+        # Temporary listens dump name.
+        req_listens_dump = config.TEMP_LISTENS_DIR
+        # req_listens_dump = self.get_req_dump(listens_dump, listens_dump_id, 2)
+        # Uncomment the above command when full and incremental dumps are updated.
+        # refer to: http://ftp.musicbrainz.org/pub/musicbrainz/listenbrainz/
+
+        self.connection.cwd(req_listens_dump)
+        listens_file_name = self.get_listens_file_name()
+
+        t0 = time()
+        current_app.logger.info('Downloading {} from FTP...'.format(listens_file_name))
+        dest_path = self.download_dump(listens_file_name, directory)
+        current_app.logger.info('Done. Total time: {:.2f} sec'.format(time() - t0))
+        return dest_path
+
+    def download_artist_relation(self, directory, artist_relation_dump_id=None):
+        """ Download artist relation to dir passed as an argument.
+
+            Args:
+                directory (str): Dir to save artist relation locally.
+                artist_relation_dump_id (int): Unique identifier of artist relation to be downloaded.
+                    If not provided, most recent artist relation will be downloaded.
+
+            Returns:
+                dest_path (str): Local path where artist relation has been downloaded.
+        """
+        self.connection.cwd('/pub/musicbrainz/listenbrainz/labs/artist-credit-artist-credit-relations/')
+        artist_relation_dump = self.list_dir()
+        req_artist_relation_dump = self.get_req_dump(artist_relation_dump, artist_relation_dump_id, 5)
```

Comment:
5? Why is this hard coded?

Reply:
Because in every dump name, the dump id is at different position

path: listenbrainz-dump-97-20191031-000002-full/
97 at path[2]

path: msid-mbid-mapping-01-20191227-121636/
01 at path[3]

path: artist-credit-artist-credit-relations-01-20191230-134806/
01 at path[5]

get_req_dump takes the dump_id_pos to get check the dump id against passed dump_id so it is constant/hard_coded for each dump (listens, mappings, artist relations)

Reply:
Ahh!